EagerExecution.swift

```diff
@@ -0,0 +1,628 @@
+// !!! THIS CODE IS AUTOMATICALLY GENERATED, DO NOT EDIT BY HAND !!!
+//
+// Copyright 2018-19 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import CTensorFlow
+
+/// **WARNING:** After constructing a `TFE_Op`, any one of its `execute` methods must be called
+/// *exactly once*. If not called, then a memory leak is introduced due to the underlying TensorFlow
+/// eager op object not being freed. If called more than once, then a SEGFAULT may occur due to
+/// trying to execute a TensorFlow eager op that has already been freed.
+@usableFromInline
+internal struct TFE_Op {
+  @usableFromInline internal let status: CTFStatus
+  @usableFromInline internal let op: CTFEOp
+
+  @inlinable @inline(__always)
+  internal init(_ name: String) {
+    self.status = TF_NewStatus()
+    self.op = TFE_NewOp(_ExecutionContext.global.eagerContext, name, status)
+  }
+
+  @inlinable @inline(__always)
+  internal func addInput(_ inputHandle: _AnyTensorHandle) -> Int {
+    TFE_OpAddInput(op, inputHandle._cTensorHandle, status)
+    guard TF_GetCode(status) == TF_OK else {
+      return 0
+    }
+    return 1
+  }
+
+  @inlinable @inline(__always)
+  internal func addInput<Scalar: TensorFlowScalar>(_ input: Tensor<Scalar>) -> Int {
+    TFE_OpAddInput(op, input.handle._cTensorHandle, status)
+    guard TF_GetCode(status) == TF_OK else {
+      return 0
+    }
+    return 1
+  }
+
+  @inlinable @inline(__always)
+  internal func addInput(_ input: StringTensor) -> Int {
+    TFE_OpAddInput(op, input.handle._cTensorHandle, status)
+    guard TF_GetCode(status) == TF_OK else {
+      return 0
+    }
+    return 1
+  }
+
+  @inlinable @inline(__always)
+  internal func addInputList<T: TensorArrayProtocol>(_ input: T) -> Int {
+    let count = input._tensorHandleCount
+    var buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: Int(count))
+    defer { buffer.deallocate() }
+    var pointer = UnsafeMutablePointer<OpaquePointer?>(buffer.baseAddress)
+    input._unpackTensorHandles(into: buffer.baseAddress)
+    TFE_OpAddInputList(op, pointer, count, status)
+    guard TF_GetCode(status) == TF_OK else {
```

Comment:
I found that this was causing test failures related to dataset. Here is the reason:
Tensorflow tries to infer the corresponding attribute that uses this length information and fails. It seems like this failure should just be ignored.

Reply:
Oh interesting. It's weird it checks for that attribute at this point and not in `TFE_OpExecute`, given that the user may set the attribute later anyway. I can remove that status check from there.

Reply:
I don't get any test failures related to datasets though. I only get some test failures related to autodiff that I haven't figured out yet and also had to add `// REQUIRES: deprecated_gpe_mode` to `dynamic_attributes.swift` as it uses `#tfop`s.

Reply:
I would think tests with tfops should not be disabled if possible. They should just be converted to raw ops.

Reply:
Sounds good. Let me look into converting those.

Reply:
The failure above is actually due to a bug in this PR. It's now fixed. The type list attribute should be set before calling `TFE_OpAddInputList`. All tests pass now.

Reply:
This is still failing for me on your new changes.

I made these changes to make it work:
35614   let op = TFE_Op("ZipDataset")                                                    
35615   op.setAttr("output_types", outputTypes)                                          
35616   op.setAttr("N", inputDatasets.count)                                             
35617   let inputDatasetsCount = op.addInputList(inputDatasets)                          
35618   op.setAttr("output_shapes", outputShapes)                                        
35619   return op.execute(Int(1)) 

Removing the check also makes this work.

Reply:
Yes sorry it requires a bit of restructuring that I'm currently doing. Removing the check makes it work but it's still wrong and could fail elsewhere later on but I know how to fix it. Is it ok if I completely remove the `tfop` mode and just make a fix for `eager` mode?

Reply:
I don't see how removing the tfop mode would help, but within this file it is ok to not generate tfop anymore. If it is to change TensorGroup, that would require compiler changes. it looks to me like eager and tfop are very different modes. What about moving all the addInput to the end after all the attributes are added? If necessary, just add a call to input._tensorHandleCount at the beginning?

Reply:
Sorry I meant to remove it so that I don't have to implement the fix for both modes. I already have a working fix that I'll push in a few minutes after I make sure tests pass.